### PR TITLE
Add enable-container-metrics feature to custom-metrics-stackdriver-adapter

### PR DESCRIPTION
**Pull request by znirzej**
https://github.com/GoogleCloudPlatform/k8s-stackdriver/issues/336

### CODE REVIEW

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 1]
```diff
@@ -103,6 +103,37 @@ func (t *Translator) GetSDReqForPods(podList *v1.PodList, metricName string, met
 	return t.createListTimeseriesRequest(joinFilters(filterForSelector, filter), metricKind), nil
 }
 
+// GetSDReqForCointainers returns Stackdriver request for query for multiple pods.
+// podList is required to be no longer than oneOfMax items. This is enforced by limitation of
+// "one_of()" operator in Stackdriver filters, see documentation:
+// https://cloud.google.com/monitoring/api/v3/filters
+func (t *Translator) GetSDReqForContainers(podList *v1.PodList, metricName string, metricKind string, metricSelector labels.Selector, namespace string) (*stackdriver.ProjectsTimeSeriesListCall, error) {
+	fmt.Println("GetSDReqForContainers", metricName, metricKind, metricSelector, namespace)
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:111]:
```suggestion
```

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 2]
```diff
@@ -540,7 +575,15 @@ func (t *Translator) getMetricValuesFromResponse(groupResource schema.GroupResou
...
-
+		containerName := series.Resource.Labels["container_name"]
+		metricContainerName, ok := metricContainer[name]
+		if !ok {
+			metricContainer[name] = containerName
+		} else {
+			if metricContainerName != containerName {
+				klog.Fatalf("Only one container in pod can have specific metric. Containers %x %x have the same metric %x in pod %x", metricContainerName, containerName, metricName, name)
+			}
+		}
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:586]: Could we move this to separate function?

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 3]
```diff
@@ -631,6 +673,8 @@ func (t *Translator) metricKey(timeSeries *stackdriver.TimeSeries) (string, erro
 		switch timeSeries.Resource.Type {
 		case "k8s_pod":
 			return timeSeries.Resource.Labels["namespace_name"] + ":" + timeSeries.Resource.Labels["pod_name"], nil
+		case "k8s_container":
+			return timeSeries.Resource.Labels["namespace_name"] + ":" + timeSeries.Resource.Labels["pod_name"], nil
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:677]: Why you selected such key that collides with pod? Would be good to document this decision by adding comment?

**Comment by znirzej** [comment]
> Looks good, just a couple of nit comments. Can you also explain how this change was tested?

I tested GetSDReqForContainers using uinit tests, unfortounatley only part of custom-metrics-stackdriver-adapter code is covered by unit tests.
I also tested changes by deploying my modifications to GCP, I used prometheus_to_sd exporter and tested HPA on it's custom metric.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 4]
```diff
@@ -366,8 +396,12 @@ func (t *Translator) filterForAnyNode() string {
 	return "resource.type = \"k8s_node\""
 }
 
+func (t *Translator) filterForAnyContainer() string {
+	return "resource.type = \"k8s_container\""
+}
+
 func (t *Translator) filterForAnyResource() string {
-	return "resource.type = one_of(\"k8s_pod\",\"k8s_node\")"
+	return "resource.type = one_of(\"k8s_pod\",\"k8s_node\",\"k8s_container\")"
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:404]: Should this be only when flag is enabled?

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 5]
```diff
@@ -521,6 +555,30 @@ func (t *Translator) createListTimeseriesRequestProject(filter string, metricKin
 		AggregationAlignmentPeriod(fmt.Sprintf("%vs", int64(alignmentPeriod.Seconds())))
 }
 
+func (t *Translator) checkContainerMetricsUniqueness(response *stackdriver.ListTimeSeriesResponse, metricName string) error {
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:558]: Please, better document what this function does or simplify it.

File: custom-metrics-stackdriver-adapter/adapter.go [hunk 6]
```diff
@@ -103,9 +106,15 @@ func main() {
 		"whether to enable Custom Metrics API")
 	flags.BoolVar(&serverOptions.EnableExternalMetricsAPI, "enable-external-metrics-api", serverOptions.EnableExternalMetricsAPI,
 		"whether to enable External Metrics API")
+	flags.BoolVar(&serverOptions.EnableContainerMetrics, "enable-container-metrics", serverOptions.EnableContainerMetrics,
```
Comment (serathius) [inline custom-metrics-stackdriver-adapter/adapter.go:109]: Please improve description of flag

**Pull request merged**